lib/spack/spack/spec.py

```diff
@@ -766,13 +767,136 @@ def __str__(self):
+    outgoing towards direct dependencies, or edges that are ingoing
+    from direct dependents.
+
+    Edges are stored in a dictionary and keyed by package name.
+    """
+    def __init__(self, store_by=EdgeDirection.child):
+        # Sanitize input arguments
+        msg = 'unexpected value for "store_by" argument'
+        assert store_by in (EdgeDirection.child, EdgeDirection.parent), msg
+
+        #: This dictionary maps a package name to a list of edges
+        #: i.e. to a list of DependencySpec objects
+        self.edges = {}
+        self.store_by_child = (store_by == EdgeDirection.child)
+
+    def __getitem__(self, key):
+        return self.edges[key]
+
+    def __iter__(self):
+        return iter(self.edges)
+
+    def __len__(self):
+        return len(self.edges)
+
+    def add(self, edge):
+        """Adds a new edge to this object.
+
+        Args:
+            edge (DependencySpec): edge to be added
+        """
+        key = edge.spec.name if self.store_by_child else edge.parent.name
+
+        current_list = self.edges.setdefault(key, [])
+
+        # The list of edges is empty, just append the edge
+        if not current_list:
+            current_list.append(edge)
+            return
+
+        # If there's something in the list, check if we need to update an
+        # already existing entry.
+        for dep in current_list:
+            if edge.spec == dep.spec and edge.parent == dep.parent:
+                dep.add_type(edge.deptypes)
+                break
+        else:
+            current_list.append(edge)
+
+        # Keep the list sorted
+        current_list.sort(key=_sort_by_dep_types)
 
     def __str__(self):
         return "{deps: %s}" % ', '.join(str(d) for d in sorted(self.values()))
 
+    def _cmp_iter(self):
+        for item in sorted(itertools.chain.from_iterable(self.edges.values())):
+            yield item
+
+    def copy(self):
+        """Copies this object and returns a clone"""
+        clone = type(self)()
+        clone.store_by_child = self.store_by_child
+
+        # Copy everything from this dict into it.
+        for dspec in itertools.chain.from_iterable(self.values()):
+            clone.add(dspec.copy())
+
+        return clone
+
+    def select_by(self, parent=None, child=None, dependency_types='all'):
+        """Return a list of edges selected by dependency types.
+
+        If an edge has any of the dependency types passed as argument
+        then it is selected.
+
+        Args:
+            parent (str): name of the parent package
+            child (str): name of the child package
+            dependency_types: allowed dependency types
+
+        Returns:
+            List of DependencySpec objects
+        """
+        if not dependency_types:
+            return []
+
+        dependency_types = dp.canonical_deptype(dependency_types)
+        # Start from all the edges we store
+        selected = [d for d in itertools.chain.from_iterable(self.values())]
```

Comment:
Use a generator instead of building the whole list here, and same for the rest of the filters
```suggestion
        selected = (d for d in itertools.chain.from_iterable(self.values()))
```

Reply:
Done in ac4a84e